File: app/Modules/Gateway/Http/Controllers/Api/TransactionController.php
<?php

namespace App\Modules\Gateway\Http\Controllers\Api;

use App\Modules\Gateway\Http\Controllers\Controller;
use App\Modules\Gateway\Models\Shop;
use App\Modules\Gateway\Models\Tag;
use App\Modules\Gateway\Models\Transaction;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;

class TransactionController extends Controller
{
    /**
     * POST /api/v1/transactions — sklep tworzy transakcję, dostaje payment_url.
     */
    public function store(Request $request): JsonResponse
    {
        /** @var Shop $shop */
        $shop = $request->attributes->get('shop');

        $data = $request->validate([
            'product_external_id' => ['required', 'string', 'max:255'],
            'product_name' => ['required', 'string', 'max:255'],
            'amount' => ['required', 'integer', 'min:1'], // grosze
            'currency' => ['nullable', 'string', 'in:PLN'],
            'return_url' => ['required', 'url', 'max:500'],
            'notify_url' => ['nullable', 'url', 'max:500'],
            'tag_uid' => ['nullable', 'string', 'max:255'],
        ]);

        $tag = null;
        if (! empty($data['tag_uid'])) {
            $tag = Tag::where('shop_id', $shop->id)->where('tag_uid', $data['tag_uid'])->first();
        }

        $transaction = Transaction::create([
            'shop_id' => $shop->id,
            'tag_id' => $tag?->id,
            'product_external_id' => $data['product_external_id'],
            'product_name' => $data['product_name'],
            'amount' => $data['amount'],
            'currency' => $data['currency'] ?? 'PLN',
            'status' => 'created',
            'mode' => $shop->payment_mode,
            'return_url' => $data['return_url'],
            'notify_url' => $data['notify_url'] ?? null,
        ]);

        return response()->json([
            'uuid' => $transaction->id,
            'payment_url' => route('pay.show', $transaction->id),
        ], 201);
    }

    /**
     * GET /api/v1/transactions/{uuid} — sklep weryfikuje status (nie ufa redirectowi).
     */
    public function show(Request $request, string $uuid): JsonResponse
    {
        /** @var Shop $shop */
        $shop = $request->attributes->get('shop');

        $transaction = Transaction::where('id', $uuid)->where('shop_id', $shop->id)->first();

        if (! $transaction) {
            return response()->json(['error' => 'Transakcja nie istnieje'], 404);
        }

        // Sklep polluje ten endpoint z ekranu "sprawdzamy płatność" —
        // przy okazji aktywnie weryfikujemy status u operatora.
        app(\App\Modules\Gateway\Services\TransactionService::class)->reconcileWithProvider($transaction);

        return response()->json([
            'uuid' => $transaction->id,
            'status' => $transaction->status,
            'amount' => $transaction->amount,
            'currency' => $transaction->currency,
            'mode' => $transaction->mode,
            'paid_at' => $transaction->paid_at?->toIso8601String(),
            'created_at' => $transaction->created_at?->toIso8601String(),
        ]);
    }
}

File: pay-backend/gateway-api/src/main/kotlin/com/supportme/pay/gateway/api/clientapi/TransactionController.kt
package com.supportme.pay.gateway.api.clientapi

import com.supportme.pay.gateway.api.payment.TransactionService
import com.supportme.pay.gateway.domain.entity.Shop
import com.supportme.pay.gateway.domain.entity.Tag
import com.supportme.pay.gateway.domain.entity.Transaction
import com.supportme.pay.gateway.domain.repository.TagRepository
import com.supportme.pay.gateway.domain.repository.TransactionRepository
import jakarta.validation.Valid
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotBlank
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.time.format.DateTimeFormatter
import java.util.UUID

data class CreateTransactionRequest(
    @field:NotBlank val productExternalId: String,
    @field:NotBlank val productName: String,
    @field:Min(1) val amount: Int,
    val currency: String? = null,
    @field:NotBlank val returnUrl: String,
    val notifyUrl: String? = null,
    val tagUid: String? = null,
)

data class CreateTransactionResponse(val uuid: String, val paymentUrl: String)

data class TransactionStatusResponse(
    val uuid: String,
    val status: String,
    val amount: Int,
    val currency: String,
    val mode: String,
    val paidAt: String?,
    val createdAt: String?,
)

/**
 * Odpowiednik `Api\TransactionController` — REST client-facing (`/api/gateway/v1/transactions`),
 * uwierzytelnione `X-Api-Key` (patrz [com.supportme.pay.gateway.api.security.ShopApiKeySecurityConfig]).
 */
@RestController
@RequestMapping("/api/gateway/v1/transactions")
class TransactionController(
    private val transactionRepository: TransactionRepository,
    private val tagRepository: TagRepository,
    private val transactionService: TransactionService,
) {

    @PostMapping
    fun store(
        @AuthenticationPrincipal shop: Shop,
        @Valid @RequestBody body: CreateTransactionRequest,
    ): ResponseEntity<CreateTransactionResponse> {
        if (body.currency != null && body.currency != "PLN") {
            return ResponseEntity.badRequest().build()
        }

        // Lookup tagu SCOPED do sklepu — po cichu ignorowany jeśli nie istnieje (jak w oryginale).
        val tag: Tag? = body.tagUid?.let { tagRepository.findByTagUidAndShop(it, shop) }

        val transaction = Transaction(
            shop = shop,
            tag = tag,
            productExternalId = body.productExternalId,
            productName = body.productName,
            amount = body.amount,
            currency = body.currency ?: "PLN",
            // mode dziedziczony ze Shop.paymentMode w momencie utworzenia — NIE wybieralny przez klienta API.
            mode = shop.paymentMode,
            returnUrl = body.returnUrl,
            notifyUrl = body.notifyUrl,
        )
        val saved = transactionRepository.save(transaction)

        return ResponseEntity.status(HttpStatus.CREATED).body(
            CreateTransactionResponse(uuid = saved.id.toString(), paymentUrl = "/pay/${saved.id}"),
        )
    }

    @GetMapping("/{uuid}")
    fun show(@AuthenticationPrincipal shop: Shop, @PathVariable uuid: UUID): ResponseEntity<Any> {
        val transaction = transactionRepository.findByIdAndShop(uuid, shop)
            ?: return ResponseEntity.status(HttpStatus.NOT_FOUND).body(mapOf("error" to "Transakcja nie istnieje"))

        // Efekt uboczny: polling tego endpointu wyzwala aktywną rekoncyliację (throttlowaną 3s).
        transactionService.reconcileWithProvider(transaction)

        return ResponseEntity.ok(
            TransactionStatusResponse(
                uuid = transaction.id.toString(),
                status = transaction.status.dbValue,
                amount = transaction.amount,
                currency = transaction.currency,
                mode = transaction.mode.dbValue,
                paidAt = transaction.paidAt?.let { DateTimeFormatter.ISO_INSTANT.format(it) },
                createdAt = transaction.createdAt?.let { DateTimeFormatter.ISO_INSTANT.format(it) },
            ),
        )
    }
}
